presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java

```diff
@@ -433,6 +465,11 @@ public void clientFailed(HttpPageBufferClient client, Throwable cause)
         {
             requireNonNull(client, "client is null");
             requireNonNull(cause, "cause is null");
+
+            // ignore failure for cancelled clients
+            if (cancelledClients.contains(client)) {
```

Comment:
Let's move this check to the `ExchangeClient.this.clientFailed` so it is under the lock.